### PR TITLE
Add records of session establishment for subscription resumption

### LINKED ISSUE #31635: [BUG] Subscription Timeout retries was broken when fixing the Readhandler creation issue
### Reproduction steps

PR #30491 Modified the Subscription Timeout logic in a such a way where the server will only retry once before deleting the persisted subscription entry.

`SubscriptionResumptionSessionEstablisher::HandleDeviceConnectionFailure` deletes the subscription entry if the server fails to establish a case session. This modifies the previously released feature without the PR mentionning this change.

To reproduce :
- Build a sample app with `chip_persist_subscriptions=true chip_subscription_timeout_resumption=true`
- Commission the device
- Establish a subscription with chip-tool **without** being in interactive mode
- Let the subscription fail since the chip-tool will not send a status response
- After 5 minutes, server will try to resub and fail
- Device will never retry to resub

### Bug prevalence

Everytime

### GitHub hash of the SDK that was being used

166c4b5eed

### Platform

core

### Platform Version(s)

_No response_

### Anything else?

This is a regression of a feature released in v1.2.

### PR DESCRIPTION
**Pull request by wqx6**
Fixes #31635

Add records of session establishment for subscription resumption and delete the subscription entry after failing to establish the session for `CHIP_CONFIG_SUBSCRIPTION_TIMEOUT_RESUMPTION_MAX_FIBONACCI_STEP_INDEX` times.

### CODE REVIEW

**Comment by yunhanw-google** [comment]
@wqx6  could you add one test to make sure we can capture the issue in https://github.com/project-chip/connectedhomeip/issues/31635? thanks

File: src/app/SubscriptionResumptionSessionEstablisher.cpp [hunk 1]
```diff
@@ -36,6 +36,32 @@ class AutoDeleteEstablisher
     SubscriptionResumptionSessionEstablisher * mSessionEstablisher;
 };
 
+struct SubscriptionResumptionSessionEstablishRecord
+{
+    chip::SubscriptionId mSubscriptionId;
+    uint16_t mSessionEstablishRetries = UINT16_MAX;
+};
+
+static SubscriptionResumptionSessionEstablishRecord sSubscriptionResumptionRecords[CHIP_IM_MAX_NUM_SUBSCRIPTIONS];
```
Comment (mkardous-silabs) [inline src/app/SubscriptionResumptionSessionEstablisher.cpp:45]: This should go in `SimpleSubscriptionResumptionStorage` and not in the establisher. 
This array isn't manage correctly. If a subscription entry is removed, this data structure is not updated.

**Review by mkardous-silabs — CHANGES_REQUESTED** [review]
PR has the right idea but does not completly fix the underlying issue.

File: src/app/SubscriptionResumptionSessionEstablisher.cpp (hunk 1 above)
Comment (wqx6) [reply src/app/SubscriptionResumptionSessionEstablisher.cpp:45]: `SimpleSubscriptionResumptionStorage` is only an example implement of `SubscriptionResumptionStorage`. Should we add the retrying logic in it?

Added updating the array when entry is removed.

Comment (wqx6) [reply src/app/SubscriptionResumptionSessionEstablisher.cpp:45]: Also I found another cause why the sever app does not retry on resuming subscriptions.

The previous work flow is:
server app tries to resume subscription and create a read handler for it-> server app fails to establish the session-> the read handler is closed and `InteractionModelEngine::OnDone()` is called->sever app schedules another subscription resumption

But the current work flow is:
sever app tries to resume subscription->server app fails to establish the session

The read handler will not be created and `InterationModelEngine::OnDone()` will not be called. So server app will never schedule another subscription resumption.

Should we schedule subscription resumption when failing to establish the session? @mkardous-silabs @jtung-apple

Comment (jtung-apple) [reply src/app/SubscriptionResumptionSessionEstablisher.cpp:45]: Subscription resumption should be scheduled when failing to establish the session. The scenario it covers is network outage. If wifi router get replaced / updated, etc, or if there's a general power outage for a day, the long-idle-time battery-powered device still should attempt to resume.

That's the reason the timeout-triggered subscription resumption is on its own timer, with the intention that that the power impact is not too much worse than regular wake+report.

After all, we want the user experience to be that after an outage of some sort, the house recovers over time without needing the user to run around the house and re-pair all the "lost" devices.

Comment (wqx6) [reply src/app/SubscriptionResumptionSessionEstablisher.cpp:45]: I have already add the change of scheduling subscription resumption when failing to establish the session. Please take a review @mkardous-silabs @jtung-apple

**Review by yunhanw-google — APPROVED** [review]

File: src/app/SubscriptionResumptionSessionEstablisher.cpp (hunk 1 above)
Comment (mkardous-silabs) [reply src/app/SubscriptionResumptionSessionEstablisher.cpp:45]: @wqx6 The intial comment was not addressed. The `SubscriptionInfo` structure is defined in the `src/app/SubscriptionResumptionStorage.h` class. Please ass this entry in that structure and not in a seperate structure.

Comment (wqx6) [reply src/app/SubscriptionResumptionSessionEstablisher.cpp]: @mkardous-silabs Done, please take a look.

File: src/app/SubscriptionResumptionSessionEstablisher.cpp [hunk 2]
```diff
@@ -109,14 +112,23 @@ void SubscriptionResumptionSessionEstablisher::HandleDeviceConnectionFailure(voi
...
+    if (subscriptionInfo.mResumptionRetries <= CHIP_CONFIG_SUBSCRIPTION_TIMEOUT_RESUMPTION_MAX_FIBONACCI_STEP_INDEX)
     {
-        subscriptionResumptionStorage->Delete(subscriptionInfo.mNodeId, subscriptionInfo.mFabricIndex,
-                                              subscriptionInfo.mSubscriptionId);
+        InteractionModelEngine::GetInstance()->TryToResumeSubscriptions();
+    }
+    else
+#endif // CHIP_CONFIG_SUBSCRIPTION_TIMEOUT_RESUMPTION
+    {
+        // If the device fails to establish the session for several times, the subscriber might be offline and its subscription
```
Comment (mkardous-silabs) [inline src/app/SubscriptionResumptionSessionEstablisher.cpp:123]:
```suggestion
        // If the device fails to establish the session several times, the subscriber might be offline and its subscription
```

File: src/app/SubscriptionResumptionSessionEstablisher.cpp [hunk 3]
```diff
@@ -109,14 +112,23 @@ void SubscriptionResumptionSessionEstablisher::HandleDeviceConnectionFailure(voi
...
+        // read client will be deleted when the device reconnect to the subscriber. This subscription will be never used again.
```
Comment (mkardous-silabs) [inline src/app/SubscriptionResumptionSessionEstablisher.cpp:124]:
```suggestion
        // read client will be deleted when the device reconnects to the subscriber. This subscription will be never used again.
```

File: src/app/SubscriptionResumptionSessionEstablisher.cpp [hunk 4]
```diff
@@ -109,14 +112,23 @@ void SubscriptionResumptionSessionEstablisher::HandleDeviceConnectionFailure(voi
...
+        // So clean up the persistent subscription information storage.
```
Comment (mkardous-silabs) [inline src/app/SubscriptionResumptionSessionEstablisher.cpp:125]:
```suggestion
        // Clean up the persistent subscription information storage.
```

**Review by mkardous-silabs — COMMENTED** [review]
I think you are missing the changes to persist the number of retries. Otherwise, the end device power cycles, will restart its retries from 0.

**Review by mkardous-silabs — APPROVED** [review]

**Pull request merged**